torch/csrc/autograd/saved_variable.h

```diff
@@ -1,46 +1,68 @@
 #pragma once
 
-#include <mutex>
-#include <memory>
-#include <functional>
-#include <ATen/ATen.h>
-
-#include "torch/csrc/jit/tracer_state.h"
-#include "torch/csrc/autograd/variable.h"
 #include "torch/csrc/autograd/variable_version.h"
-#include "torch/csrc/Types.h"
 
-namespace torch { namespace autograd {
+#include <ATen/Tensor.h>
+
+#include <cstdint>
+#include <list>
+#include <memory>
 
+namespace torch {
+namespace autograd {
+class Variable;
 struct Function;
+} // namespace autograd
+namespace jit { namespace tracer {
+struct ValueTracingStateElem;
+using ValueTracingState = std::list<ValueTracingStateElem>;
+}} // namespace jit::tracer
+} // namespace torch
 
-extern const char* ERR_BACKWARD_TWICE;
+namespace torch { namespace autograd {
 
-struct SavedVariable {
-  SavedVariable()
-    : data()
-    , has_grad_fn(false)
-    , version()
-    , requires_grad(false)
-    , expected_version(-1) {}
+extern const char* ERR_BACKWARD_TWICE;
 
+/// A snapshot of a variable at a certain version. A `SavedVariable` stores
+/// enough information to reconstruct a variable from a certain point in time.
+class SavedVariable {
+ public:
+  SavedVariable() = default;
   SavedVariable(const Variable& variable, bool is_output);
+  SavedVariable(SavedVariable&&) = default;
+  SavedVariable& operator=(SavedVariable&&) = default;
+
+  // Must be defined externally to avoid it being inlined by the compiler,
+  // which would require it to see the definition of ValueTracingState.
+  ~SavedVariable();
+
+  /// Reconstructs the saved variable. Pass `saved_for` as the gradient
+  /// function if constructing the `SavedVariable` with it would have caused a
+  /// circular reference.
+  Variable unpack(std::shared_ptr<Function> saved_for = nullptr) const;
+
+  void reset_data() {
+    return data_.reset();
+  }
+
+ private:
+  bool was_default_constructed_ = true;
+  at::Tensor data_;
 
-  at::Tensor data;
   // The gradient function associated with this node. If has_grad_fn
   // is false, then this is a leaf node. Note that the grad_fn is not saved if
   // it would create a circular reference. In that case, the grad_fn must be
   // passed in to the unpack function when reconstructing the Variable.
-  bool has_grad_fn;
-  std::shared_ptr<Function> _grad_fn;
-  std::weak_ptr<Function> grad_accumulator;
-  SavedVersion version;
-  bool requires_grad;
-  int expected_version;
-  int output_nr;
-  std::unique_ptr<jit::tracer::ValueTracingState> tracing_state;
-
-  Variable unpack(std::shared_ptr<Function> saved_for=nullptr) const;
-};
+  std::shared_ptr<Function> grad_fn_;
+  bool has_grad_fn_ = false;
```

Comment:
Can you compress those bools in one place? Right now the struct will likely have 1B for a bool, 7B for padding, 8B for grad_fn, 1B for bool, 7B for padding, ...

Reply:
Sounds like a good idea, will implement.